sources/api/pluto/src/main.rs

```diff
@@ -327,60 +315,74 @@ async fn get_private_dns_name(client: &mut ImdsClient) -> Result<String> {
-}
-
 async fn run() -> Result<()> {
-    let setting_name = parse_args(env::args());
     let mut client = ImdsClient::new();
+    let aws_k8s_info = api::get_aws_k8s_info().await.context(error::AwsInfoSnafu)?;
 
-    let setting = match setting_name.as_ref() {
-        "cluster-dns-ip" => get_cluster_dns_ip(&mut client).await,
-        "node-ip" => get_node_ip(&mut client).await,
-        // If we want to specify a reasonable default in a template, we can exit 2 to tell
-        // sundog to skip this setting.
-        "max-pods" => get_max_pods(&mut client)
-            .await
-            .map_err(|_| process::exit(2)),
-        "provider-id" => get_provider_id(&mut client).await,
-        "private-dns-name" => get_private_dns_name(&mut client).await,
-        _ => usage(),
-    }?;
-
-    // sundog expects JSON-serialized output so that many types can be represented, allowing the
-    // API model to use more accurate types.
-
-    // 'max_pods' setting is an unsigned integer, convert 'settings' to u32 before serializing to JSON
-    if setting_name == "max-pods" {
-        let max_pods = serde_json::to_string(
-            &setting
-                .parse::<u32>()
-                .context(error::ParseToU32Snafu { setting: &setting })?,
-        )
-        .context(error::OutputJsonSnafu { output: &setting })?;
-        println!("{}", max_pods);
+    // Generate the cluster-dns-ip
+    let cluster_dns_ip = if let Some(ip) = aws_k8s_info
+        .cluster_dns_ip
+        .as_ref()
+        .and_then(|x| x.iter().next())
+    {
+        *ip
     } else {
-        let output =
-            serde_json::to_string(&setting).context(error::OutputJsonSnafu { output: &setting })?;
-        println!("{}", output);
-    }
+        let ip = get_cluster_dns_ip(&mut client, &aws_k8s_info).await?;
+        IpAddr::from_str(&ip).context(error::BadIpSnafu { ip: ip.clone() })?
+    };
+
+    // Generate the node-ip
+    let node_ip = if let Some(node_ip) = aws_k8s_info.node_ip.as_ref() {
+        node_ip.clone()
+    } else {
+        get_node_ip(&mut client, cluster_dns_ip).await?
+    };
+
+    // Generate max-pods
+    // Skip the set if this fails
+    let max_pods = if let Some(max_pods) = aws_k8s_info.max_pods.as_ref() {
+        Some(*max_pods)
+    } else if let Ok(max_pods) = get_max_pods(&mut client).await {
+        Some(max_pods)
+    } else {
+        None
+    };
+
+    // Generate provider-id
+    let provider_id = if let Some(provider_id) = aws_k8s_info.provider_id.as_ref() {
+        provider_id.clone()
+    } else {
+        get_provider_id(&mut client).await?
+    };
+
+    // Generate private-dns-name
+    let private_dns_name = if NO_HOSTNAME_VARIANTS.contains(&aws_k8s_info.variant_id.as_str()) {
+        None
+    } else if let Some(hostname_override) = aws_k8s_info.hostname_override {
+        Some(hostname_override.clone())
+    } else {
+        Some(get_private_dns_name(&mut client, &aws_k8s_info).await?)
+    };
```

Comment:
These chained `if` statements make it hard to follow which values we actually have and when.

Refactoring all of this along these lines could make it more clear and be more straightforward to extend in the future:
```
generate_cluster_dns_ip(&mut client, &mut aws_k8s_info).await?;
generate_node_ip(&mut client, &mut aws_k8s_info).await?;
generate_max_pods(&mut client, &mut aws_k8s_info).await?;
generate_private_dns_name(&mut client, &mut aws_k8s_info).await?;
```

Even better would be to keep `aws_info` and `k8s_info` separate, since we don't expect to mutate the AWS-related fields. Then when you're done calling the generate functions, you can serialize the k8s struct and return that.

There's an ordering dependency where we must run `generate_cluster_dns_ip` first, before we run `generate_node_ip`. We could enforce that by having `generate_node_ip` call `generate_cluster_dns_ip` right away - if it does an early check / return for an existing value, that will be quick, and then the main function doesn't have to risk getting it wrong.